source/extensions/clusters/aggregate/cluster.h

```diff
@@ -37,6 +40,24 @@ class Cluster : public Upstream::ClusterImplBase, Upstream::ClusterUpdateCallbac
           Server::Configuration::TransportSocketFactoryContextImpl& factory_context,
           Stats::ScopePtr&& stats_scope, ThreadLocal::SlotAllocator& tls, bool added_via_api);
 
+  struct PerThreadLoadBalancer : public ThreadLocal::ThreadLocalObject {
+    AggregateClusterLoadBalancer& get() {
+      // We can refresh before the per-worker LB is created. One of these variants should hold
+      // a non-null value.
+      if (absl::holds_alternative<std::unique_ptr<AggregateClusterLoadBalancer>>(lb_)) {
+        ASSERT(absl::get<std::unique_ptr<AggregateClusterLoadBalancer>>(lb_) != nullptr);
+        return *absl::get<std::unique_ptr<AggregateClusterLoadBalancer>>(lb_);
+      } else {
+        ASSERT(absl::get<AggregateClusterLoadBalancer*>(lb_) != nullptr);
+        return *absl::get<AggregateClusterLoadBalancer*>(lb_);
+      }
+    }
+
+    // For aggregate cluster the per-thread LB is only created once. We need to own it so we
+    // can pre-populate it before the LB is created and handed to the cluster.
+    absl::variant<std::unique_ptr<AggregateClusterLoadBalancer>, AggregateClusterLoadBalancer*> lb_;
```

Comment:
Hmm, not very happy with this "maybe owned" semantics here. In my debug yesterday seems we may delay the call to `refresh` [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/clusters/aggregate/cluster.cc#L85) to make sure LB is created and handed before thread local cluster is updated.

Reply:
I thought about this a fair amount. The issue is there is no dependency order between clusters so we have to be very careful to not lose updates. I think it may be possible to move all of the logic to the thread local cluster updates (off of the main thread) and initialize first during LB creation on each worker. My opinion though is we should merge this since it will work and people are complaining about this and I can circle back. But up to you. I can try to refactor further.

Reply:
I looked into this a little more and it's pretty tricky because the thread local load balancer is created in the constructor for the thread local cluster, so the cluster won't exist at that point. I can probably make this better but it's not easy and I recommend going with this for now. Let me know what you think.

Reply:
Actually, sorry, nevermind, we don't need to look up ourself, just other clusters. Let me see what I can do.

Reply:
I've spent a bit of time trying to implement the alternate version and it's not so easy. I will see what I can do but I would still recommend we go with this for now and I will try to replace it with something better.

Reply:
OK, yeah agreed it is tricky. I'm ok with going with this for now. Do we want backport this?

Reply:
@lizan I opened https://github.com/envoyproxy/envoy/pull/14388 which is a trivial fix and will be easy to backport. The more complicated fix here is required because of the unrelated changes in cluster_manager_impl. I will work on a better version of this PR in parallel and it will not need to be backported.

/wait